docs: add ARO Classic AI debugging and log-search hints (#4836)
Azure/ARO-RP: docs/ai/classic-log-search.md

# Agentic Hints for Log Search in ARO Classic

- IMPORTANT: this document is referenced by agentic workflows, DO NOT REMOVE IT.
- Read `docs/ai/classic-debugging.md` first. Load this file only when the current task is primarily log search or request tracing from logs.
- This file assumes a Kusto/Geneva log-search skill. Do not assume access to admin APIs, `oc`, Prometheus, or live cluster actions.
- If any of the info below turns out not to be accurate, suggest an update PR at the end of the session.

> ## Differences from ARO HCP

Reviewer: **Edit 3 — insert a mandatory "Validate Before You Trust" section, plus an explicit "Column Mappings" table.**

This is the most important change. The doc currently gives an agent field names without a way to confirm those names exist in the schema being queried, so a zero-row result from a misspelled or differently-cased column will be reported as "no events" — a confident false negative. Forcing a schema check before any conclusion closes that gap.

It also surfaces environment-specific column meanings (e.g. `Tenant` = region in ARO Classic Geneva) as positive facts the agent must use, not warnings to heed. The agent cannot derive `Tenant == region` from the column name, so the doc has to state it.

The suggestion below inserts the new section above `## Differences from ARO HCP` by replacing that header with `<new section> + <original header>`:

```suggestion
## Validate Before You Trust

Before any query whose result will drive a conclusion, you MUST:

1. Resolve and record the concrete Kusto cluster, database, and table being queried. If these are not known from the workspace or session, consult the ARO SRE TSGs under "Authoritative References." Do not guess table names.
2. Run a schema check on each table before filtering (`<table> | getschema` or `.show table <T> schema`). Record which referenced fields are top-level columns and which live inside a `Properties` / dynamic bag. Adjust the query shape accordingly (e.g. `tostring(Properties.request_id)` instead of `request_id`).
3. Treat a zero-row result as ambiguous until all of the following are true:
   - the table contains rows in the time window (unfiltered `count` on the same window),
   - every referenced column exists in `getschema`,
   - a relaxed version of the filter (dropping the most specific predicate) also returns zero.
   Only then conclude "no matching events." Otherwise report the query as inconclusive and explain which check failed.
4. Field names listed elsewhere in this document are a starting hypothesis, not ground truth. If a name does not appear in `getschema`, search the equivalent property in the dynamic bag before concluding the data is missing. For columns whose meaning is not obvious from the name, use the explicit mapping in "Column Mappings" below or in the linked TSG — never infer meaning from the column name.

## Column Mappings

These are environment-specific facts the agent cannot derive from column names alone. Use them by reference.

- `Tenant` in the ARO Classic Geneva tables maps to the Azure region. Use it as the region filter. Do not assume this generalizes to other Geneva deployments — confirm per environment via the linked TSG.
- `operation_id` is RP-generated. It is not the `Azure-AsyncOperation` polling ID embedded in the polling URL.
- `resource_id`, `subscription_id`, `resource_group`, `resource_name`, `cluster_deployment_namespace` may be lowercased by enrichment but preserved as-is when derived from request paths. Match case-insensitively using `=~` or `tolower()` on both sides.
- Add new mappings here as they are discovered, rather than embedding them in prose elsewhere in this doc.

## Differences from ARO HCP
```

> - Do not assume the HCP Kusto tables (`HostedControlPlaneLogs`, `frontendLogs`, `backendLogs`, etc.) or Grafana datasource names exist in ARO Classic.

Reviewer 2: Instead of listing specific HCP table names to avoid (which may become stale), consider directing agents to verify available tables against the actual Kusto/Geneva DB schema as the single source of truth.

> - The repo does not document a complete Classic-specific Kusto table inventory. Start with cross-family pivots like `request_id` and often `resource_id`, then narrow with family-specific fields such as `request_path`, `LOGKIND`, and known log messages instead of HCP-specific table names.

Reviewer: **Edit 4 — replace the "no inventory exists" footnote with an authoritative pointer.**

Acknowledging the gap doesn't close it. Point at the TSGs as the source of record for table identity and make clear that cross-family pivots like `request_id` narrow *within* a chosen table, not in place of choosing one.

```suggestion
- The authoritative ARO Classic Kusto table inventory lives in the ARO SRE TSGs (see "Authoritative References"). Resolve the concrete cluster, database, and table from there before querying. Cross-family pivots like `request_id` and `resource_id` and family-specific fields like `request_path`, `LOGKIND`, and known log messages are how you narrow *within* a chosen table — they do not substitute for choosing the right table.
```

- ARO Classic logs are emitted through Geneva/MDSD. Metrics flow through MDM/statsd, but metrics are out of scope for this file.

## Start Here

- If you already have an RP response or log line for the original RP-handled request, start with `request_id`.
- If you are looking at an `Azure-AsyncOperation` polling request, pivot on the async operation ID from the polling URL/path first; each poll gets a fresh `request_id`.
- If you do not have `request_id`, start with `resource_id`; fall back to `subscription_id`, `resource_group`, `resource_name`, and a time window.
- Customer `PUT` / `PATCH` / `DELETE` cluster mutations are async. Default search order: frontend access logs -> `frontendqos` -> backend `asyncqos` -> outbound dependency logs -> optional Hive / installer logs.
- For admin or maintenance incidents that are already visible in logs, use admin route names only as `request_path` filters. These are often frontend-only log trails, so start with `request_path` plus handler-specific messages and do not assume `frontendqos` or backend `asyncqos` unless code inspection shows an async handoff. This skill cannot call those endpoints.

## High-Value Fields

> - `request_id` is the best primary pivot for a single RP-handled request, and usually for the original mutation request. The RP always generates it and returns it in `X-Ms-Request-Id`.

Reviewer: **Edit 7 — call out emitter-side vs searchable column naming once, at the top of "High-Value Fields".**

Every field in this section is the *emitter-side* name from the ARO-RP source. The searchable Kusto column can differ in casing or live inside a `Properties` / dynamic bag. Stating this once at the top of the section, with a pointer to the validation rule, prevents the agent from filtering on a name that doesn't exist as a top-level column.

```suggestion
- The field names in this section are emitter-side names from the ARO-RP source. The searchable column name and casing in Kusto may differ (e.g. `request_id` may surface as a top-level column, a `Properties.request_id` entry, or under a different case). Confirm via `getschema` on the target table before filtering. See "Validate Before You Trust."
- `request_id` is the best primary pivot for a single RP-handled request, and usually for the original mutation request. The RP always generates it and returns it in `X-Ms-Request-Id`.
```

- `correlation_id` and `client_request_id` come from caller or ARM headers and may be empty; use them opportunistically.
- `operation_id` is also generated by the RP and appears in logs, but do not confuse it with the `Azure-AsyncOperation` polling URL.
- For `Azure-AsyncOperation` status/result polling, use the async operation ID embedded in the URL/path to bridge back to the original operation; each poll has a new `request_id`.
- Inbound logs enrich from request paths, so matching paths typically produce `subscription_id`, `resource_group`, `resource_name`, and `resource_id`.
- Backend logs enrich from the cluster resource ID and can also include `cluster_deployment_namespace`, which is useful for bridging into Hive-backed flows.

> - Prefer case-insensitive matching for `resource_id`, `subscription_id`, `resource_group`, `resource_name`, and `cluster_deployment_namespace`; some enrichment lowercases these fields while path-derived values can preserve incoming case.

Reviewer: **Edit 6 — make the case-insensitivity rule concrete.**

"Prefer case-insensitive matching" is too soft for an LLM agent — it will default to `==`. Spell out the operators so the rule is mechanical.

```suggestion
- Match `resource_id`, `subscription_id`, `resource_group`, `resource_name`, and `cluster_deployment_namespace` case-insensitively, using `=~` or `tolower()` on both sides. Enrichment lowercases these fields in some paths and preserves incoming case in others, so `==` will produce false negatives.
```

## Log Families

- Access logs from `pkg/frontend/middleware/log.go` use the messages `read request` and `sent response`. They carry `request_method`, `request_path`, `request_user_agent`, and, on `sent response`, `response_status_code` and `duration`.
- `LOGKIND=frontendqos` in `pkg/frontend/frontend.go` marks ARM-facing operation-result logs with `operationType`, `resultType`, and, on failures, `errorDetails`. Most admin routes do not use this wrapper; for those, access logs and handler-specific log lines are the main search surface.
- `LOGKIND=asyncqos` in `pkg/backend/openshiftcluster.go` marks terminal backend results with the messages `long running operation succeeded` and `long running operation failed`. These logs add `operationType`, `resultType`, `clusterIdentity`, and, on failure, `errorDetails`.
- `HttpRequestStart` and `HttpRequestEnd` in `pkg/util/azureclient/roundtripper.go` add `LOGKIND=outboundRequests` when full outbound logging is enabled. In local dev, this path is disabled by default unless `ARO_ENABLE_OUTBOUND_HTTP_LOGGING=true`.

> - `OutboundRequestFailed` and `OutboundRequestConflict` from the same logging policy are always important search terms, even when `LOGKIND=outboundRequests` is absent. They include fields like `request_host`, `request_method`, and `response_status_code`.

Reviewer 2: This mentions outbound HTTP logging is disabled in local dev unless `ARO_ENABLE_OUTBOUND_HTTP_LOGGING=true`, but it might be worth noting that in production it is always enabled unconditionally — see `outboundHTTPLoggingEnabled()` in `roundtripper.go` line 157: when not in development mode it returns `true` immediately.

- If a control-plane resize incident is already suspected, useful messages from `pkg/frontend/admin_openshiftcluster_resize_controlplane.go` include `Resizing control plane node`, `Cordoning node`, `Draining node`, `Stopping VM`, `Resizing VM`, `Starting VM`, `Waiting for node`, `Uncordoning node`, `Updating Machine object`, and `Updating Node labels`.

## Code Locations

- `pkg/api/correlation.go` - correlation field generation
- `pkg/util/log/log.go` - resource and correlation field enrichment
- `pkg/frontend/middleware/log.go` - inbound request/access log fields
- `pkg/frontend/frontend.go` - `frontendqos` operation-result logs
- `pkg/frontend/openshiftcluster_putorpatch.go` - async handoff for create and update
- `pkg/frontend/openshiftcluster_delete.go` - async handoff for delete
- `pkg/backend/openshiftcluster.go` - `asyncqos` terminal backend logs
- `pkg/util/azureclient/roundtripper.go` - outbound dependency logging
- `pkg/util/log/hive.go` - Hive / installer log-field propagation

> - `pkg/operator/controllers/genevalogging/` - cluster log shipping via `fluentbit` and `mdsd`

Reviewer: **Edit 5 — append an "Authoritative References" section at the end, with placeholders per log family.**

Pairs with Edit 1 and Edit 4: this is where the per-log-family TSG links live, so the rest of the doc can defer to it. Filling the placeholders with concrete TSG URLs can be a follow-up — the structure itself is what makes the rest of the doc grounded instead of name-only.

```suggestion
- `pkg/operator/controllers/genevalogging/` - cluster log shipping via `fluentbit` and `mdsd`

## Authoritative References

- This document is authoritative for: which ARO-RP code path emits which log message and which fields are added there.
- The ARO SRE TSGs on `eng.ms` are authoritative for: Kusto cluster / database / table names, validated query templates per log family, environment-specific column meanings (e.g. `Tenant`), and current Geneva account/namespace mappings.
- TSG links per log family (fill in and date-stamp as they are confirmed):
  - Access logs (`read request` / `sent response`): <TSG URL> (read YYYY-MM-DD)
  - `LOGKIND=frontendqos`: <TSG URL> (read YYYY-MM-DD)
  - `LOGKIND=asyncqos`: <TSG URL> (read YYYY-MM-DD)
  - `LOGKIND=outboundRequests` / `OutboundRequestFailed` / `OutboundRequestConflict`: <TSG URL> (read YYYY-MM-DD)
  - Hive / installer log propagation: <TSG URL> (read YYYY-MM-DD)
  - Cluster log shipping (`mdsd` / Geneva logging controller): <TSG URL> (read YYYY-MM-DD)
- When a TSG link disagrees with a field name or message string described above, open an update PR against this doc.
```
